#16b990586  ch(databaseSetup manipulatingData)  setup database, versioning and make requests to endpoints

banobepascal opened the pull request:
#### What does this PR do?
- Versions our Api with v2
- setup database to store, manipulate, retrieve dat
- sets endpoints for database

#### Description of Task to be completed?
- completed user signin and signup
- completed endpoint GET all articles posted
- completed endpoint for GET specific article with ID
- completed endpoint for POST an article
- completed enpoint for EDIT an article
- completed endpoint for DELETE article
- completed endpoint for GET  specific articles with tag or title
- completed enpoint for flagging an article as inappropiate

#### How should this be manually tested?
A user can navigate to the branch(ch-databaseSetup-16b990586) and clone the repo then run `npm run dbSetup` to create database

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/1689905867

banobepascal committed:
ch(database setup)

- setting up database to store,
- manipulate and retrieve data
- using npm package pg to connect to postgres
- versioning of api with v2

[Finishes #168990586]

NkFab commented on api/v2/controllers/adminController.js:
```diff
@@ -0,0 +1,25 @@
+import articles from '../models/article';
+
+class AdminAccess {
+  // delete specific flagged article
```
There's a better way to comment on methods using Jsdocs format. For instance, you can do rewrite this comment as /**
            * @description delete specific flagged article
            * @param {Object} req
            * @param {Object} res
            * @returns {Object} Returns the response
            / ***

*Note:* Do this on every method where you have comments.

banobepascal replied:
thanks for the recomendation point noted

NkFab commented on api/v2/controllers/articleController.js:
```diff
@@ -0,0 +1,146 @@
+      });
+    }
+
+    const { error } = validation.validateFlag(req.body);
```
A method should only do what it was created for, in this case, it was created to flag an article and nothing else, so you should move the validation in its separate folder and apply it in the route as a middleware.

banobepascal replied:
working on it

NkFab commented on api/v2/controllers/commentController.js:
```diff
@@ -0,0 +1,45 @@
+      });
+    }
+
+    const { error } = validation.validateComment(req.body);
```
A method should only do what it was created for, in this case, it was created to create a comment on an article and nothing else, so you should move the validation in its separate folder and apply it in the route as a middleware.

NkFab commented on api/v2/controllers/userController.js:
```diff
@@ -0,0 +1,101 @@
+class UserController {
+  // create user account
+  static async signUp(req, res) {
+    const { error } = validation.validateUserSignUp(req.body);
```
A method should only do what it was created for, in this case, it was created to create an account and nothing else, so you should move the validation in its separate folder and apply it in the route as a middleware.

NkFab commented on api/v2/controllers/userController.js:
```diff
@@ -0,0 +1,101 @@
+      });
+    }
+
+    const validPassword = await bcrypt.compare(
```
This should be moved to a separate helper folder for encrypting passcodes.

NkFab commented on api/v2/controllers/userController.js:
```diff
@@ -0,0 +1,101 @@
+      isAdmin: checkEmail.isAdmin,
+    };
+
+    const gentoken = jwt.sign(signinPayLoad, process.env.JWT_KEY);
```
The token generation should be moved to a helper function directory.

NkFab commented on api/v2/middleware/admin.js:
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable func-names */
+/* eslint-disable consistent-return */
+
+module.exports = function (req, res, next) {
```
Why are you using es5?

NkFab commented on api/v2/middleware/prod.js:
```diff
@@ -0,0 +1,8 @@
+import compression from 'compression';
+
+// eslint-disable-next-line func-names
+module.exports = function (app) {
```
Why are you mixing es5 with es6?

NkFab commented on api/v2/migrations/dbSetup.js:
```diff
@@ -0,0 +1,23 @@
+  port: 5432,
+  max: 10,
+  idleTimeoutMillis: 30000,
+};
```
This config object data shouldn't be accessible to the public move it to the dotenv file.

NkFab commented on api/v2/migrations/dropTables.js:
```diff
@@ -0,0 +1,23 @@
+  port: 5432,
+  max: 10,
+  idleTimeoutMillis: 30000,
+};
```
This config object data shouldn't be accessible to the public move it to the dotenv file.

NkFab commented on api/v2/models/article.js:
```diff
@@ -0,0 +1,44 @@
+  },
+];
+
+export default articles;
```
Is this file still needed? If yes, for what?

NkFab commented on api/v2/models/comment.js:
```diff
@@ -0,0 +1,18 @@
+  },
+];
+
+export default comments;
```
Is this file still needed? If yes, for what?

banobepascal replied:
thanks for the feedback i am working on it

NkFab commented on api/v2/models/users.js:
```diff
@@ -0,0 +1,27 @@
+  },
+];
+
+export default users;
```
Is this file still needed? If yes, for what?

NkFab commented on api/v2/routes/admin.js:
```diff
@@ -0,0 +1,14 @@
+
+adminRoute.delete(
+  '/api/v1/flagged/articles/:id',
+  [validation.checkToken, adminAuth],
```
Why did you add the brackets on these imported middlewares?

banobepascal committed:
ch(database setup):

- refactored files in v2
- removed model file
- created view all articles
- created get specific article by id
- created flag article as innapropiate
- created edit article by id
- created delete article by id
- created user signup
- created post articles/share articles

[Finishes #16b990586]

banobepascal committed:
ch(databaseSetup-168990586)

- added user login to database
- a user can view all articles
- a user can view a specific article
- a user can edit an article
- a user can delete an article
- a user can flag an article inappropiate
- a user can view an article with a specific tag or title

[Finishes #168990586]

banobepascal committed:
ch(database setup):

-  completed user signin and signup
- - completed endpoint GET all articles posted
- completed endpoint for GET specific article with ID
- completed endpoint for POST an article
- completed enpoint for EDIT an article
- completed endpoint for DELETE article
- completed endpoint for GET  specific articles with tag or title
- completed enpoint for flagging an article as inappropiate

[Finishes #168990586]

NkFab commented on .env:
```diff
@@ -1 +1,3 @@
-JWT_KEY=teamwork
\ No newline at end of file
+JWT_KEY=teamwork
+
+DATABASE_URL=postgres://postgres:postgres@localhost:5432/teamworkdb
```
The .env file shouldn't available to the public because there's no need to have it if it will be available. Git-ignore this file.

NkFab commented on api/v2/middleware/verifyToken.js:
```diff
@@ -0,0 +1,39 @@
+/* eslint-disable consistent-return */
+import ENV from 'dotenv';
+import jwt from 'jsonwebtoken';
+// import client from '../helpers/dbConnection';
```
Remove this commented line.

NkFab reviewed (commented):
Increase your code coverage

banobepascal commented:
Hey Manzi i am going to resolve this PR and send you a new PR in 10 minutes with the appropiate git branch convention. since it is affects my Tracker stories
